hardware/arduino/sam/cores/arduino/wiring_digital.c

```diff
@@ -90,6 +114,14 @@ extern void digitalWrite( uint32_t ulPin, uint32_t ulVal )
 
 extern int digitalRead( uint32_t ulPin )
 {
+  if ((g_pinStatus[ulPin] & 0xF) == PIN_STATUS_DIGITAL_OUTPUT) {
+    return (g_pinStatus[ulPin] & 0xF0) >> 4;
+  }
+
+  if ((g_pinStatus[ulPin] & 0xF) == PIN_STATUS_ANALOG) {
+    pinMode(ulPin, INPUT);
```

Comment:
Indeed, the main goal of this patchset was solving the incompatibilities between pins behaviour on AVR and on SAM core, losing as few performances as we can.
For sure this piece of code can be optimized but, as you said, it's really uncommon to use `pinMode` during the loop, so we can be quite sure that it won't affect most people.